### PR TITLE
Move About to a widget

### LINKED ISSUE #696: Feedback on About feature
I find the api for about a bit odd. Imho it should be a widget. I don't understand why it is defined in the Application trait, since it's not something that libcosmic can position automatically. It would be more consistent to just call `widget::about(About {..})`. I don't really see an use case for overwridding `fn about()`.

@edfloreshz any thought ? i can work on it if its ok

### PR DESCRIPTION
**Pull request by edfloreshz**
Per @wiiznokes suggestion in #696, the About section code was moved to a widget.

Closes #696

### CODE REVIEW

**Review by mmstick — APPROVED** [review]

**Comment by mmstick** [comment]
Some tests are failing with an unresolved import for `about`

**Comment by edfloreshz** [comment]
Sorry, should be fixed.

File: src/widget/mod.rs [hunk 1]
```diff
@@ -365,3 +365,9 @@ pub use warning::*;
 #[cfg(feature = "markdown")]
 #[doc(inline)]
 pub use iced::widget::markdown;
+
+#[cfg(feature = "desktop")]
+pub mod about;
+#[cfg(feature = "desktop")]
+#[doc(inline)]
+pub use about::{about, About};
```
Comment (wiiznokes) [inline src/widget/mod.rs:373]:
```suggestion
pub use about::about;
```

File: src/widget/about.rs [hunk 2]
```diff
@@ -117,3 +125,81 @@ impl About {
         self
     }
 }
+
+#[cfg(feature = "desktop")]
```
Comment (wiiznokes) [inline src/widget/about.rs:129]: this can be removed

File: src/widget/about.rs [hunk 3]
```diff
@@ -1,9 +1,17 @@
...
+use {
+    crate::{
+        iced::{alignment::Vertical, Alignment, Length},
+        widget::{self, horizontal_space},
+        Element,
+    },
+    std::collections::BTreeMap,
+};
 
 #[cfg(feature = "desktop")]
```
Comment (wiiznokes) [inline src/widget/about.rs:11]: this can be removed

File: src/widget/mod.rs (hunk 1 above)
Comment (wiiznokes) [reply src/widget/mod.rs:373]: nevermind this, it's not a widget

Comment (edfloreshz) [reply src/widget/mod.rs]: I actually think this is a good idea.

File: src/widget/about.rs (hunk 2 above)
Comment (edfloreshz) [reply src/widget/about.rs]: Removed!

File: src/widget/about.rs (hunk 3 above)
Comment (edfloreshz) [reply src/widget/about.rs]: Removed!

**Comment by wiiznokes** [comment]
some nitpicks: 
- would be nice if the fields were private, and the `set_` prefix were removed.
- some data are duplicated. For example, repository_url is stored in a dedicated field and the links BtreeMap. I can see the value of having dedicated fields, for example, to add custom icons in the future
- would be nice to add a method `add_link(self, name, link) -> Self`
- `set_support_url` is ambigus imo, for example, i have this value set in metainfo.xml: `bugtracker`, `donation`, `vcs-browser`, `homepage`
- maybe switch `license` with `license_type`, and rename `license_type` to `licence_text`

wdyt @edfloreshz  ?

**Comment by edfloreshz** [comment]
Those are great suggestions, I'll make the changes in the next couple of hours.

**Comment by wiiznokes** [comment]
here are some ideas of urls we can add https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#url

**Comment by edfloreshz** [comment]
- Properties are now private.
- Removed the URL specific methods and made method `links` public, letting anyone pass an array of label/link pairs.
- Added the `license` crate to handle licenses. 
- Removed `license_type` and replaced it with a link to the license page (https://spdx.org/licenses).

@wiiznokes

File: src/widget/about.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    icon: Option<String>,
+    /// The application’s version.
+    version: Option<String>,
+    /// Name of the application's author.
+    author: Option<String>,
+    /// Comments about the application.
+    comments: Option<String>,
+    /// The application's copyright.
+    copyright: Option<String>,
+    /// The license text.
```
Comment (wiiznokes) [inline src/widget/about.rs:29]:
```suggestion
    /// The license name.
```

File: src/widget/about.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    license: Option<String>,
+    /// Artists who contributed to the application.
```
Comment (wiiznokes) [inline src/widget/about.rs:31]: Documentation for fields with `#[setters(skip)]` should be moved to the setter function instead.

File: src/widget/about.rs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            .map(|(k, v)| (k.to_string(), format!("mailto:{v}")))
+            .collect();
+        self
+    }
+
+    pub fn developers(mut self, developers: impl Into<BTreeMap<&'a str, &'a str>>) -> Self {
+        let developers: BTreeMap<&'a str, &'a str> = developers.into();
+        self.developers = developers
+            .into_iter()
+            .map(|(k, v)| (k.to_string(), format!("mailto:{v}")))
```
Comment (wiiznokes) [inline src/widget/about.rs:83]: Could we provide a way to link something else than a email address ? Or just nothing. (maybe by checking if an `@` is present, or if the string is empty)

**Review by wiiznokes — APPROVED** [review]
I like the new api!
Just some minor doc issues

File: src/widget/about.rs (hunk 6 above)
Comment (edfloreshz) [reply src/widget/about.rs:83]: Those fields are meant for email addresses, if the user needs to provide a URL they can with `links`.

**Comment by wiiznokes** [comment]
Could you add this method to support the `fl!` macro ?

```rs
pub fn link(mut self, name: impl Into<String>, link: &'a str) -> Self {
    self.links.insert(name.into(), link.to_string());
    self
}
```

**Comment by edfloreshz** [comment]
I'm thinking of the best way to support localization with the existing set of methods.

**Comment by wiiznokes** [comment]
> I'm thinking of the best way to support localization with the existing set of methods.

smtg like that could work:

```rs
pub fn links<const N: usize>(
            mut self,
            links: [(impl Into<String>, impl Into<String>); N],
        ) -> Self {
            self.links = links
                .into_iter()
                .map(|(k, v)| (k.into(), v.into()))
                .collect();
            self
        }
```

**Comment by edfloreshz** [comment]
Decided to do this instead:

```rust
pub fn links<T: Into<String>>(mut self, links: impl Into<Vec<(T, &'a str)>>) -> Self {
    let links: Vec<(T, &'a str)> = links.into();
    self.links = links
        .into_iter()
        .map(|(k, v)| (k.into(), v.to_string()))
        .collect();
    self
}
```

**Comment by wiiznokes** [comment]
Yeah, it works. The `Vec` also make probably more sense.

I noticed that the scrollbar is a bit off, but idk if it's related or not. Maybe we need to adjust the padding.

Capture d’écran du 2024-11-09 21-08-58

**Comment by edfloreshz** [comment]
Yeah we could give it some right padding

**Comment by git-f0x** [comment]
Is there a reason that the scrollable was removed from the drawer? Since now it isn't possible to make the scrollbar be at the right edge (to match designs), and that spacing required to make the scrollbar not cover content makes things off-center.

**Comment by wiiznokes** [comment]
> @wiiznokes Please confirm if I can ask Michael to review.

Yeah, go ahead

**Comment by git-f0x** [comment]
If #698 gets merged, the scrollbar spacing can be removed, but `.padding([0, space_l, space_l, space_l])` should be applied to the column (before the scrollable).

**Review by mmstick — APPROVED** [review]

**Pull request merged**